net/interfaces/interfaces_windows.go

```diff
@@ -7,76 +7,91 @@ package interfaces
-         10.0.0.0      255.255.0.0         On-link        10.0.28.63    261
-       10.0.28.63  255.255.255.255         On-link        10.0.28.63    261
-        10.0.42.0    255.255.255.0   100.103.42.106   100.103.42.106      5
-     10.0.255.255  255.255.255.255         On-link        10.0.28.63    261
-   34.193.248.174  255.255.255.255   100.103.42.106   100.103.42.106      5
-
-*/
 func likelyHomeRouterIPWindows() (ret netaddr.IP, ok bool) {
-	cmd := exec.Command("route", "print", "-4")
-	cmd.SysProcAttr = &syscall.SysProcAttr{HideWindow: true}
-	stdout, err := cmd.StdoutPipe()
+	rs, err := winipcfg.GetIPForwardTable2(windows.AF_INET)
 	if err != nil {
+		// TODO: It would be nice to log this...
 		return
 	}
-	if err := cmd.Start(); err != nil {
-		return
+
+	// Effective metric is sum of interface metric and route metric offset
+	setEffectiveMetric := func(r *winipcfg.MibIPforwardRow2) bool {
+		iface, err := r.InterfaceLUID.IPInterface(windows.AF_INET)
+		if err != nil {
+			// TODO: It would be nice to log this...
+			return false
+		}
+		r.Metric += iface.Metric
+		return true
 	}
-	defer cmd.Wait()
 
-	var f []mem.RO
-	lineread.Reader(stdout, func(lineb []byte) error {
-		line := mem.B(lineb)
-		if !mem.Contains(line, mem.S("0.0.0.0")) {
-			return nil
+	unspec := net.IPv4(0, 0, 0, 0)
+	var route *winipcfg.MibIPforwardRow2
+	metricUpdated := false
+
+	for _, r := range rs {
+		if r.Loopback || r.DestinationPrefix.PrefixLength != 0 || !r.DestinationPrefix.Prefix.IP().Equal(unspec) {
+			// Not a default route, so skip
+			continue
+		}
+
+		ip, ok := netaddr.FromStdIP(r.NextHop.IP())
+		if !ok || ip.IsZero() {
+			// Not a valid gateway, so skip (won't happen though)
+			continue
+		}
+
+		if route == nil {
+			route = &r
+			ret = ip
+			continue
 		}
-		f = mem.AppendFields(f[:0], line)
-		if len(f) < 3 || !f[0].EqualString("0.0.0.0") || !f[1].EqualString("0.0.0.0") {
-			return nil
+
+		// We can get here only if there are multiple default gateways defined (rare case),
+		// in which case we need to calculate the effective metric.
+		// First update the previous route metric if not already updated:
+		if !metricUpdated {
+			if !setEffectiveMetric(route) {
+				// TODO: What to do if the interface metric cannot be loaded? Maybe just ignore?
+				return netaddr.IP{}, false
+			}
+			metricUpdated = true
+		}
+
+		// Update current route metric:
+		if !setEffectiveMetric(&r) {
+			// TODO: What to do if the interface metric cannot be loaded? Maybe just ignore?
+			return netaddr.IP{}, false
 		}
-		ipm := f[2]
-		ip, err := netaddr.ParseIP(string(mem.Append(nil, ipm)))
-		if err == nil && isPrivateIP(ip) {
+
+		if route.Metric > r.Metric || route.Metric == r.Metric && ret.Compare(ip) > 0 {
+			// Pick the route with lower metric, or lower IP if metrics are equal
+			route = &r
 			ret = ip
 		}
-		return nil
-	})
+	}
+
+	if !ret.IsZero() && !isPrivateIP(ret) {
```

Comment:
in retrospect, this `isPrivateIP` check was probably misguided. I think I only added it because I wasn't confident of the parsing before and we didn't care about accuracy much, as we were only using it for stats collection at the time (to see which port mapping protocols were available in the wild). But I don't think we need this check now.

Reply:
But if we remove `isPrivateIP` check then the function will return the gateway even if it isn't private (even if the machine is directly connected to the internet and uses ISP's gateway), which contradicts the function name (`likelyHomeRouterIPWindows`) that suggests it's **home** router.

It's easy for me to remove the check, but need you to confirm again that you're aware of that.

Reply:
Okay, keep it for now. I can remove it later if/when I remove the "Home" part from the name. I need to talk to @danderson about that first.